### PR TITLE
Update alsa packages

### PR DESCRIPTION
**Pull request by L-as**
###### Motivation for this change

A bunch of the alsa packages were quite out of date.
Some have had attempted updates before but have been reverted:
- #126413 
- #126462

In addition, many of the previous patches and customisations have been
removed as they are no longer necessary.

Cross-compilation for alsa-firmware does still work.

###### Things done

I've tested this on my Acer Chromebook Spin 513 and it works flawlessly.

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by L-as** [comment]
The reason this is one PR is that some of the versions likely depend on each other, e.g. alsa-ucm-conf on alsa-lib

**Comment by L-as** [comment]
Discussion in https://github.com/NixOS/nixpkgs/issues/21751 is relevant, albeit in the case of `alsa-firmware`, the output seemingly doesn't actually differ depending on the platform (tested with `diff -r`). I'm not sure what it's compiling, but it will probably work fine even on when cross-compiled.

**Comment by collares** [comment]
I don't know if this problem is a blocker, but this breaks cross-compilation, as can be seen in the logs below. If you're on x86_64 and compiling for aarch64, as in the example below, CC should output aarch64 binaries but CC_FOR_BUILD should output x86_64 binaries. You might have something like `boot.binfmt.emulatedSystems = [ "aarch64-linux" ];` in your config masking the mismatch.

<details>
  <summary>Build log for pkgsCross.aarch64-multiplatform.alsa-firmware on x86_64.</summary>
```
building '/nix/store/r5kgb4zkw82ax5qlqn3acyc5j06gn3a7-alsa-firmware-aarch64-unknown-linux-gnu-1.2.4.drv'...
unpacking sources
unpacking source archive /nix/store/z4sarpvzn4rv9a1c6i895l0fxvskvr6i-alsa-firmware-1.2.4.tar.bz2
source root is alsa-firmware-1.2.4
setting SOURCE_DATE_EPOCH to timestamp 1603193562 of file alsa-firmware-1.2.4/configure
patching sources
autoreconfPhase
autoreconf: export WARNINGS=
autoreconf: Entering directory '.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: aclocal --force
autoreconf: configure.ac: tracing
autoreconf: configure.ac: not using Libtool
autoreconf: configure.ac: not using Intltool
autoreconf: configure.ac: not using Gtkdoc
autoreconf: running: /nix/store/x5q70zjlassqa65m616cqdra18yvxl9z-autoconf-2.71/bin/autoconf --force
configure.ac:5: warning: The macro `AC_PROG_CC_FOR_BUILD' is obsolete.
configure.ac:5: You should run autoupdate.
m4/ax_prog_cc_for_build.m4:54: AC_PROG_CC_FOR_BUILD is expanded from...
configure.ac:5: the top level
configure.ac:8: warning: The macro `AC_HEADER_STDC' is obsolete.
configure.ac:8: You should run autoupdate.
./lib/autoconf/headers.m4:704: AC_HEADER_STDC is expanded from...
configure.ac:8: the top level
configure.ac:56: warning: AC_OUTPUT should be used without arguments.
configure.ac:56: You should run autoupdate.
autoreconf: configure.ac: not using Autoheader
autoreconf: running: automake --add-missing --copy --force-missing
vxloader/Makefile.am:47: warning: '%'-style pattern rules are a GNU make extension
autoreconf: './config.sub' is updated
autoreconf: './config.guess' is updated
autoreconf: Leaving directory '.'
updateAutotoolsGnuConfigScriptsPhase
Updating Autotools / GNU config script to a newer upstream version: ./config.sub
Updating Autotools / GNU config script to a newer upstream version: ./config.guess
configuring
fixing libtool script ./ltmain.sh
configure flags: --disable-dependency-tracking --prefix=/nix/store/qncbv68s88c0wkspm0vc3hsrnfymjnsw-alsa-firmware-aarch64-unknown-linux-gnu-1.2.4 --with-hotplug-dir=\$\(out\)/lib/firmware --build=x86_64-unknown-linux-gnu --host=aarch64-unknown-linux-gnu
checking for aarch64-unknown-linux-gnu-gcc... aarch64-unknown-linux-gnu-gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... yes
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether aarch64-unknown-linux-gnu-gcc accepts -g... yes
checking for aarch64-unknown-linux-gnu-gcc option to enable C11 features... none needed
checking whether aarch64-unknown-linux-gnu-gcc understands -c and -o together... yes
checking how to run the C preprocessor... aarch64-unknown-linux-gnu-gcc -E
checking build system type... x86_64-unknown-linux-gnu
checking for x86_64-unknown-linux-gnu-gcc... aarch64-unknown-linux-gnu-gcc
checking whether we are using the GNU C compiler... yes
checking whether aarch64-unknown-linux-gnu-gcc accepts -g... yes
checking for aarch64-unknown-linux-gnu-gcc option to enable C11 features... (cached) none needed
checking whether aarch64-unknown-linux-gnu-gcc understands -c and -o together... yes
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... configure: error: in `/build/alsa-firmware-1.2.4':
configure: error: cannot run C compiled programs.
If you meant to cross compile, use `--build'.
See `config.log' for more details
error: builder for '/nix/store/r5kgb4zkw82ax5qlqn3acyc5j06gn3a7-alsa-firmware-aarch64-unknown-linux-gnu-1.2.4.drv' failed with exit code 77;
       last 10 log lines:
       > checking whether aarch64-unknown-linux-gnu-gcc accepts -g... yes
       > checking for aarch64-unknown-linux-gnu-gcc option to enable C11 features... (cached) none needed
       > checking whether aarch64-unknown-linux-gnu-gcc understands -c and -o together... yes
       > checking whether the C compiler works... yes
       > checking for C compiler default output file name... a.out
       > checking for suffix of executables...
       > checking whether we are cross compiling... configure: error: in `/build/alsa-firmware-1.2.4':
       > configure: error: cannot run C compiled programs.
       > If you meant to cross compile, use `--build'.
       > See `config.log' for more details
       For full logs, run 'nix log /nix/store/r5kgb4zkw82ax5qlqn3acyc5j06gn3a7-alsa-firmware-aarch64-unknown-linux-gnu-1.2.4.drv'.
```
</details>

**Comment by collares** [comment]
I guess you can do `export CC_FOR_BUILD=$CC` only if this is *not* a cross build, and cross builds would be handled normally by autoconf.

**Comment by collares** [comment]
I guess you can set `CC_FOR_BUILD` to `${buildPackages.stdenv.cc}/bin/cc`.

File: pkgs/os-specific/linux/alsa-project/alsa-firmware/default.nix [hunk 1]
```diff
@@ -1,20 +1,19 @@
...
-    sha256 = "1aq8z8ajpjvcx7bwhwp36bh5idzximyn77ygk3ifs0my3mbpr8mf";
+    sha256 = "sha256-tnttfQi8/CR+9v8KuIqZwYgwWjz1euLf0LzZpbNs1bs=";
   };
 
-  patches = [ (fetchpatch {
-    url = "https://github.com/alsa-project/alsa-firmware/commit/a8a478485a999ff9e4a8d8098107d3b946b70288.patch";
-    sha256 = "0zd7vrgz00hn02va5bkv7qj2395a1rl6f8jq1mwbryxs7hiysb78";
-  }) ];
+  postPatch = ''
+    export CC_FOR_BUILD=$CC
```
Comment (sternenseemann) [inline pkgs/os-specific/linux/alsa-project/alsa-firmware/default.nix:13]: This likely breaks cross compilation if it actually ever uses `CC_FOR_BUILD` for something (which is the only way I could imagine this working now). Try instead:

```
  depsBuildBuild = [ buildPackages.stdenv.cc ];
```

which sets `CC_FOR_BUILD` already.

**Comment by L-as** [comment]
@collares 
> You might have something like boot.binfmt.emulatedSystems = [ "aarch64-linux" ]; in your config masking the mismatch.

This is definitely the issue. So much for purity!

File: pkgs/os-specific/linux/alsa-project/alsa-firmware/default.nix (hunk 1 above)
Comment (L-as) [reply pkgs/os-specific/linux/alsa-project/alsa-firmware/default.nix]: This seems to have fixed it, thanks!

**Comment by collares** [comment]
Should this target staging?

**Comment by L-as** [comment]
#151645 #151649 #151646 don't target staging, so I think it's fine.

**Comment by Artturin** [comment]
More than 500 rebuilds so this should target staging 
The bit generated ucm update also does.

**Comment by L-as** [comment]
Fixed.

**Review by Artturin — APPROVED** [review]
tested compiling alsa-utils and `pkgsCross.aarch64-multiplatform.alsa-utils` but couldn't test running because of (i assume) mismatched system lib versions
```
./result/bin/speaker-test -t wav -c 6

speaker-test 1.2.6

Playback device is default
Stream parameters are 48000Hz, S16_LE, 6 channels
WAV file(s)
ALSA lib pcm.c:2576:(snd_pcm_open_conf) Unknown field libs
Playback open error: -22,Invalid argument
```

**Comment by L-as** [comment]
FWIW you can try the new ALSA for your system while avoiding a whole system recompilation by using Firefox with the old ALSA, that was the main culprit in my case. Firefox does still work, though crashes occasionally when you do this.

**Comment by Artturin** [comment]
the firefox test does test audio so i'll run that https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/firefox.nix

**Comment by L-as** [comment]
I assume the tests succeded. I've been using this version of ALSA for over a week now without any problems.

**Comment by L-as** [comment]
bump

File: pkgs/os-specific/linux/alsa-project/alsa-firmware/default.nix [hunk 2]
```diff
@@ -1,20 +1,16 @@
...
+    sha256 = "sha256-tnttfQi8/CR+9v8KuIqZwYgwWjz1euLf0LzZpbNs1bs=";
   };
 
-  patches = [ (fetchpatch {
-    url = "https://github.com/alsa-project/alsa-firmware/commit/a8a478485a999ff9e4a8d8098107d3b946b70288.patch";
-    sha256 = "0zd7vrgz00hn02va5bkv7qj2395a1rl6f8jq1mwbryxs7hiysb78";
-  }) ];
-
-  nativeBuildInputs = [ autoreconfHook buildPackages.stdenv.cc ];
+  depsBuildBuild = [ buildPackages.stdenv.cc ];
```
Comment (AndersonTorres) [inline pkgs/os-specific/linux/alsa-project/alsa-firmware/default.nix:12]: what is the difference between buildPackages.stdenv and stdenv itself?

File: pkgs/os-specific/linux/alsa-project/alsa-firmware/default.nix [hunk 3]
```diff
@@ -32,10 +28,11 @@ stdenv.mkDerivation rec {
...
 
-  meta = {
+  meta = with lib; {
     homepage = "http://www.alsa-project.org/";
     description = "Soundcard firmwares from the alsa project";
-    license = lib.licenses.gpl2Plus;
-    platforms = lib.platforms.linux;
+    license = licenses.gpl2Plus;
+    platforms = platforms.linux;
+    maintainers = [ maintainers.l-as ];
```
Comment (AndersonTorres) [inline pkgs/os-specific/linux/alsa-project/alsa-firmware/default.nix:36]:
```suggestion
    maintainers = with maintainers; [ l-as ];
    platforms = platforms.linux;
```

File: pkgs/os-specific/linux/alsa-project/alsa-lib/default.nix [hunk 4]
```diff
@@ -7,24 +7,15 @@
 
 stdenv.mkDerivation rec {
   pname = "alsa-lib";
-  version = "1.2.5.1";
+  version = "1.2.6.1";
 
   src = fetchurl {
     url = "mirror://alsa/lib/${pname}-${version}.tar.bz2";
-    sha256 = "sha256-YoQh2VDOyvI03j+JnVIMCmkjMTyWStdR/6wIHfMxQ44=";
+    sha256 = "rVgpk9Us21+xWaC+q2CmrFfqsMwb34XcTbbWGX8CMz8=";
```
Comment (AndersonTorres) [inline pkgs/os-specific/linux/alsa-project/alsa-lib/default.nix:14]:
```suggestion
    hash = "sha256-rVgpk9Us21+xWaC+q2CmrFfqsMwb34XcTbbWGX8CMz8=";
```

File: pkgs/os-specific/linux/alsa-project/alsa-ucm-conf/default.nix [hunk 5]
```diff
@@ -30,7 +30,7 @@ stdenv.mkDerivation rec {
     '';
 
     license = licenses.bsd3;
-    maintainers = [ maintainers.roastiek ];
+    maintainers = [ maintainers.roastiek maintainers.l-as ];
```
Comment (AndersonTorres) [inline pkgs/os-specific/linux/alsa-project/alsa-ucm-conf/default.nix:33]:
```suggestion
    maintainers = with maintainers; [ roastiek l-as ];
```

File: pkgs/os-specific/linux/alsa-project/alsa-ucm-conf/default.nix [hunk 6]
```diff
@@ -1,12 +1,12 @@
...
-  name = "alsa-ucm-conf-${version}";
-  version = "1.2.5.1";
+  pname = "alsa-ucm-conf";
+  version = "1.2.6.3";
 
   src = fetchurl {
-    url = "mirror://alsa/lib/${name}.tar.bz2";
-    sha256 = "sha256-WEGkRBZty/R523UTA9vDVW9oUIWsfgDwyed1VnYZXZc=";
+    url = "mirror://alsa/lib/${pname}-${version}.tar.bz2";
+    sha256 = "uKA6o4emJKL2XtwgG/d3QhGQtgUpqSCHZGgjr72Wxc0=";
```
Comment (AndersonTorres) [inline pkgs/os-specific/linux/alsa-project/alsa-ucm-conf/default.nix:9]:
```suggestion
    hash = "sha256-uKA6o4emJKL2XtwgG/d3QhGQtgUpqSCHZGgjr72Wxc0=";
```

File: pkgs/os-specific/linux/alsa-project/alsa-plugins/default.nix [hunk 7]
```diff
@@ -21,7 +21,7 @@ stdenv.mkDerivation rec {
     description = "Various plugins for ALSA";
     homepage = "http://alsa-project.org/";
     license = licenses.lgpl21;
-    maintainers = [ maintainers.marcweber ];
+    maintainers = [ maintainers.marcweber maintainers.l-as ];
```
Comment (AndersonTorres) [inline pkgs/os-specific/linux/alsa-project/alsa-plugins/default.nix:24]:
```suggestion
    maintainers = with maintainers; [ marcweber l-as ];
```

File: pkgs/os-specific/linux/alsa-project/alsa-lib/default.nix [hunk 8]
```diff
@@ -43,5 +34,6 @@ stdenv.mkDerivation rec {
 
     license = licenses.lgpl21Plus;
     platforms = platforms.linux;
+    maintainers = [ maintainers.l-as ];
```
Comment (AndersonTorres) [inline pkgs/os-specific/linux/alsa-project/alsa-lib/default.nix:37]:
```suggestion
    maintainers = with maintainers; [ l-as ];
```

**Review by AndersonTorres — DISMISSED** [review]

File: pkgs/os-specific/linux/alsa-project/alsa-lib/default.nix (hunk 4 above)
Comment (Artturin) [reply pkgs/os-specific/linux/alsa-project/alsa-lib/default.nix:14]: *
```suggestion
    sha256 = "sha256-rVgpk9Us21+xWaC+q2CmrFfqsMwb34XcTbbWGX8CMz8=";
```

File: pkgs/os-specific/linux/alsa-project/alsa-plugins/default.nix [hunk 9]
```diff
@@ -2,11 +2,11 @@
 
 stdenv.mkDerivation rec {
   pname = "alsa-plugins";
-  version = "1.2.5";
+  version = "1.2.6";
 
   src = fetchurl {
     url = "mirror://alsa/plugins/${pname}-${version}.tar.bz2";
-    sha256 = "086z2g2f95570vfvp9d5bakib4k18fb4bszf3lgx3j6j6f2gkvj2";
+    sha256 = "BogYpLVdjAKdqgABXYU9RRE/VrIkt8ZOHhF5iMglsqA=";
```
Comment (AndersonTorres) [inline pkgs/os-specific/linux/alsa-project/alsa-plugins/default.nix:9]:
```suggestion
    hash = "sha256-BogYpLVdjAKdqgABXYU9RRE/VrIkt8ZOHhF5iMglsqA=";
```

File: pkgs/os-specific/linux/alsa-project/alsa-firmware/default.nix (hunk 2 above)
Comment (Artturin) [reply pkgs/os-specific/linux/alsa-project/alsa-firmware/default.nix:12]: if a compiler is needed for `buildplatform -> buildplatform` then 
`depsBuildBuild = [ buildPackages.stdenv.cc ]; ` is used

**Comment by L-as** [comment]
Is there some rule that says we need to change `sha256 = ...` to `hash = ...`?

**Comment by AndersonTorres** [comment]
> Is there some rule that says we need to change `sha256 = ...` to `hash = ...`?

Shortest answer:

No.

**Comment by L-as** [comment]
I do not feel like working on this PR anymore.

**Comment by L-as** [comment]
In the future I suggest that you don't merge PRs like #154038 that cause merge conflicts with this one and don't offer anything more.

**Comment by bobby285271** [comment]
Shall I revert #154038 instead (https://github.com/NixOS/nixpkgs/pull/154038#issuecomment-1008744224)?

**Comment by bobby285271** [comment]
Well actually I still prefer rebasing this to master, do you mind do that?

**Comment by L-as** [comment]
> Well actually I still prefer rebasing this to master, do you mind do that?

Sure, I rebased @bobby285271

**Comment by L-as** [comment]
Ah, just noticed you merged #151649 . Only alsa-firmware here hasn't been updated. I'll close this again.

**Comment by bobby285271** [comment]
Maybe also `alsa-lib`? (oh that should target staging...)

**Pull request closed without merging**